### PR TITLE
[#357] Add basic notifications

### LINKED ISSUE #357: Support basic notification types
We should implement notification functionality start with:
* s3:TestEvent
* s3:ObjectCreated:* 
* s3:ObjectRemoved:*

Blocked by #340

### LINKED ISSUE #358: Add ObjectCreated events
* s3:ObjectCreated:Put
* s3:ObjectCreated:Post
* s3:ObjectCreated:Copy
* s3:ObjectCreated:CompleteMultipartUpload 

Also check existance of   __s3:ObjectCreated:*__  in configuration and send notifications in all cases above

Don't send notifications from failed operations.

s3:ObjectCreated:CompleteMultipartUpload includes objects that are created using UploadPartCopy for Copy operations.

### LINKED ISSUE #359: Add objectDeleted events
* s3:ObjectRemoved:Delete -- notification when an object is deleted or a versioned object is permanently deleted by using the s3:ObjectRemoved:Delete event type.
* s3:ObjectRemoved:DeleteMarkerCreated -- when a delete marker is created for a versioned object 

Also add checking of existence of __s3:ObjectRemoved:*__ in notification configurations, which means sending notifications in all cases above. 

These event notifications don't alert for automatic deletes from Lifecycle policies or from failed operations.

### PR DESCRIPTION
**Pull request by masterSplinter01**
Closes #357, closes #358, closes #359

### CODE REVIEW

File: api/notifications/controller.go [hunk 1]
```diff
@@ -118,3 +248,200 @@ func (c *Controller) Listen(ctx context.Context) {
 		}(stream)
 	}
 }
+
+// CheckBucketConfiguration checks notification configuration and generates ID for configurations with empty ids.
+func (c *Controller) CheckBucketConfiguration(conf *data.NotificationConfiguration, r *api.ReqInfo) (completed bool, err error) {
```
[the remark this answers is not in the document]
Comment (KirillovDenis) [reply api/notifications/controller.go:253]: In general I think we shouldn't couple the `Controller` (that handle common nats pub/sub logic) with S3 notification logic (checking configuration, filtering s3 events and so on)

File: api/handler/multipart_upload.go [hunk 2]
```diff
@@ -426,7 +427,28 @@ func (h *handler) CompleteMultipartUploadHandler(w http.ResponseWriter, r *http.
...
+	if err := h.obj.SendNotifications(r.Context(), s); err != nil {
+		h.log.Error("couldn't send notification: %w", zap.Error(err))
+	}
+
+	bktSettings, err := h.obj.GetBucketSettings(r.Context(), bktInfo)
+	if err != nil {
+		h.logAndSendError(w, "could not get bucket settings", reqInfo, err)
+		return
+	}
+
```
Comment (KirillovDenis) [inline api/handler/multipart_upload.go:445]: Maybe we should use this settings below to determine if we need set `api.AmzVersionID` header (lines 462-466)

File: api/handler/delete.go [hunk 3]
```diff
@@ -83,6 +98,43 @@ func (h *handler) DeleteObjectHandler(w http.ResponseWriter, r *http.Request) {
...
+				Name:    reqInfo.ObjectName,
+				HashSum: deletedObject.DeleteMarkerEtag,
+			},
+			BktInfo: bktInfo,
+			ReqInfo: reqInfo,
+		}
+
+		if err := h.obj.SendNotifications(r.Context(), m); err != nil {
+			h.log.Error("couldn't send notification: %w", zap.Error(err))
+		}
```
Comment (KirillovDenis) [inline api/handler/delete.go:114]: We can move this from the if branch to avoid duplication

File: api/notifications/controller.go [hunk 4]
```diff
@@ -114,3 +180,83 @@ func (c *Controller) Listen(ctx context.Context) {
 		}(stream)
 	}
 }
+
+func (c *Controller) SendNotifications(topics map[string]string, p *layer.SendNotificationParams) error {
+	event := prepareEvent(p.Event, p.User, p.BktInfo, p.ObjInfo, p.ReqInfo)
```
Comment (KirillovDenis) [inline api/notifications/controller.go:185]: Why we don't provide `p` as argument?

File: api/layer/layer.go [hunk 5]
```diff
@@ -532,21 +543,25 @@ func (n *layer) CopyObject(ctx context.Context, p *CopyObjectParams) (*data.Obje
...
 	var (
 		err error
 		ids []*oid.ID
+
+		versioningEnabled = false
 	)
 
+	if settings != nil {
+		versioningEnabled = settings.VersioningEnabled
+	}
```
Comment (KirillovDenis) [inline api/layer/layer.go:556]: After https://github.com/nspcc-dev/neofs-s3-gw/pull/390 settings cannot be nil

File: api/handler/multipart_upload.go [hunk 6]
```diff
@@ -437,9 +457,7 @@ func (h *handler) CompleteMultipartUploadHandler(w http.ResponseWriter, r *http.
 		Key:    objInfo.Name,
 	}
 
-	if settings, err := h.obj.GetBucketSettings(r.Context(), bktInfo); err != nil {
-		h.log.Warn("couldn't get bucket versioning", zap.String("bucket name", reqInfo.BucketName), zap.Error(err))
-	} else if settings.VersioningEnabled {
+	if bktSettings != nil && bktSettings.VersioningEnabled {
```
Comment (KirillovDenis) [inline api/handler/multipart_upload.go:460]: After https://github.com/nspcc-dev/neofs-s3-gw/pull/390 settings cannot be nil

**Review by alexvanin — APPROVED** [review]
LGTM. As we discussed with @masterSplinter01 and @KirillovDenis, it is possible to improve internal structure by decoupling some components. E.g.:
- define different interfaces for producing and handling notifications,
- implement those interfaces in controller component,
- move notification production from `layer` to `handler` with interface from step 1.
- keep notification handling in `layer` with interface from step 1.

**Pull request merged**